checker/src/main/java/org/checkerframework/checker/mustcall/MustCallTransfer.java

```diff
@@ -0,0 +1,375 @@
+    String targetStrWithoutAdaptation =
+        AnnotationUtils.getElementValue(
+            createsObligation, supplier.getCreatesObligationValueElement(), String.class, "this");
+    // TODO: find a way to also check if the target is a known tempvar, and if so return that. That
+    // should improve the quality of the error messages we give.
+    JavaExpression targetExpr;
+    try {
+      targetExpr =
+          StringToJavaExpression.atMethodInvocation(
+              targetStrWithoutAdaptation, n, atypeFactory.getChecker());
+      if (targetExpr instanceof Unknown) {
+        issueUnparseableError(n, atypeFactory, targetStrWithoutAdaptation);
+        return null;
+      }
+    } catch (JavaExpressionParseException e) {
+      issueUnparseableError(n, atypeFactory, targetStrWithoutAdaptation);
+      return null;
+    }
+    return targetExpr;
+  }
+
+  /**
+   * Issues a mustcall.not.parseable error.
+   *
+   * @param n the node
+   * @param atypeFactory the type factory to use to issue the error
+   * @param targetStrWithoutAdaptation the unparseable string
+   */
+  private static void issueUnparseableError(
+      MethodInvocationNode n,
+      GenericAnnotatedTypeFactory<?, ?, ?, ?> atypeFactory,
+      String targetStrWithoutAdaptation) {
+    atypeFactory
+        .getChecker()
+        .reportError(
+            n.getTree(),
+            "mustcall.not.parseable",
+            n.getTarget().getMethod().getSimpleName(),
+            targetStrWithoutAdaptation);
+  }
+
+  /**
+   * This method either creates or looks up the temp var t for node, and then updates the store to
+   * give t the same type as node.
+   *
+   * @param node the node to be assigned to a temporal variable
+   * @param result the transfer result containing the store to be modified
+   */
+  public void updateStoreWithTempVar(TransferResult<CFValue, CFStore> result, Node node) {
+    // Must-call obligations on primitives are not supported.
+    if (!TypesUtils.isPrimitiveOrBoxed(node.getType())) {
+      LocalVariableNode temp = getOrCreateTempVar(node);
+      if (temp != null) {
+        JavaExpression localExp = JavaExpression.fromNode(temp);
+        AnnotationMirror anm =
+            atypeFactory
+                .getAnnotatedType(node.getTree())
+                .getAnnotationInHierarchy(atypeFactory.TOP);
+        insertIntoStores(result, localExp, anm == null ? atypeFactory.TOP : anm);
+      }
+    }
+  }
+
+  /**
+   * Either returns the temporary variable associated with node, or creates one if one does not
+   * exist.
+   *
+   * @param node a node, which must be an expression (not a statement)
+   * @return a temporary variable node representing {@code node} that can be placed into a store
+   */
+  private @Nullable LocalVariableNode getOrCreateTempVar(Node node) {
+    LocalVariableNode localVariableNode = atypeFactory.tempVars.get(node.getTree());
+    if (localVariableNode == null) {
```

Comment:
`localVariableNode` can be `null` even if the lookup succeeded, because line 327 below runs even if `localVariableNode` is null.  It seems this should check `containsKey` before doing new computation.

Reply:
I don't think so. I think it was an error to unconditionally update the result in the cache - it should only have been updated when `localVariableNode` was null - i.e. when new computation was necessary.